src/adic_space.lean

```diff
@@ -8,7 +8,128 @@ import for_mathlib.topology
 
 open nat function
 
-variables (R : Type) [comm_ring R] [topological_space R] [topological_ring R]  
+section comm_ring
+variables (R : Type) [comm_ring R]
+-- This section is filled in in Johan's PR
+definition is_subring {R : Type} [comm_ring R] : set R → Prop := sorry 
+definition is_integrally_closed {R : Type} [comm_ring R] : set R → Prop := sorry  
+end comm_ring
+
+section topological_ring
+
+variables (R : Type*) [comm_ring R] [topological_space R] [topological_ring R]  
+
+def nhd_zero := (nhds (0 : R)).sets
+variable {R}
+lemma nhd_zero_symmetric {V : set R} : V ∈ nhd_zero R →  (λ a, -a) '' V ∈ nhd_zero R :=
+begin
+  intro H,
+  have := continuous.tendsto (topological_add_group.continuous_neg R) 0,
+  rw (show (λ (a : R), -a) 0 = 0, by simp) at this,
+  have almost:= this H,
+  have aux : { r : R | -r ∈ V } = (λ a, -a) '' V, by simp[set.image, neg_eq_iff_neg_eq],
+  simpa [filter.mem_map, aux] using almost
+end
+
+-- peredicates we need for topological rings
+
+-- Following is copy-pasted from t2_space class.
+-- We need to think whether we could directly use that class
+definition is_hausdorff (α : Type*) [topological_space α] : Prop := 
+  ∀x y, x ≠ y → ∃u v : set α, is_open u ∧ is_open v ∧ x ∈ u ∧ y ∈ v ∧ u ∩ v = ∅
+
+
+open filter lattice
+lemma half_nhd {α : Type*} [topological_space α] [add_group α] [topological_add_group α] (U ∈ (nhds (0 : α)).sets) : 
+  ∃ V ∈ (nhds (0 : α)).sets, ∀ v w ∈ V, v + w ∈ U :=
+/- Here is the story : by continuity of addition, and because 0 + 0 = 0,
+   (+)⁻¹(U) is a neighborhood of (0, 0). Hence it contains some V₁ × V₂. Then set V = V₁ ∩ V₂ -/
+begin
+  have nhdb_in_prod : ((λ a : α × α, a.1+a.2)⁻¹' U) ∈ (nhds ((0, 0) : α × α)).sets,
+    by apply tendsto_add' ; simp [H],
+  rw nhds_prod_eq at nhdb_in_prod,
+  rcases mem_prod_iff.1 nhdb_in_prod with ⟨V₁, H₁, V₂, H₂, H⟩,
+  have H12: V₁ ∩ V₂ ∈ (nhds (0 : α)).sets := inter_mem_sets H₁ H₂,    
+  existsi [(V₁ ∩ V₂), H12],
+  intros v w Hv Hw,
+  have : (v,w) ∈ set.prod V₁ V₂, by finish,
+  simpa using H this
+end
+
+
+instance toplogical_ring.to_uniform_space : uniform_space R := 
+uniform_space.of_core { 
+  uniformity  := (⨅ U ∈ nhd_zero R, principal {p : R×R | p.2 - p.1 ∈ U}),
```

Comment:
It says in the docstring of uniform_space that a topological group has a uniform structure. Do you think this might already be proved in mathlib? If so, we could just apply it to the underlying additive group of R, right?